Port `synth_cz_depth_line_mr` to Rust

jlofti opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR is aims to advance #12212 by porting #12229 specifically. 

### Details and comments

This PR ports the `synth_cz_depth_line_mr` function to Rust. Notes:

- I added the port in the `lnn.rs` file similar to how `pmh.rs` is currently implemented. 
- Following the advice of using gate sequences as a return type then building circuits from `CircuitData`.
- Made a little bit of optimizations by removing an inner j and patcnt loop in `synth_cz_depth_line_mr`
- Broke out shared logic into `lnn::_append_phase_gate`
- Added `_append_cx_stage1/2` in the `lnn.rs` file, but can break them out if we want them in seperate file more similar to Python

Please let me know any changes to make and I'll take care of it

Thanks!

fixes #12229

jlofti committed:
First pass, pre cleanup

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
optimizations and cleanup

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
more cleanup

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
broke out function

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
reformat

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
comments for append_cx

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=12949) <br/>All committers have signed the CLA.

jlofti committed:
Merge branch 'main' into port-synth-cz-depth-line-mr

jlofti committed:
correct usize usage

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
Merge branch 'port-synth-cz-depth-line-mr' of https://github.com/jlofti/qiskit into port-synth-cz-depth-line-mr

ShellyGarion commented on crates/accelerate/src/synthesis/linear/mod.rs:
```diff
@@ -175,6 +177,14 @@ fn check_invertible_binary_matrix(py: Python, mat: PyReadonlyArray2<bool>) -> Py
+#[pyo3(signature = (mat))]
+fn synth_cz_depth_line_mr(py: Python, mat: PyReadonlyArray2<bool>) -> PyResult<CircuitData> {
+    let view = mat.as_array();
+    let (num_qubits, lnn_gates) = lnn::synth_cz_depth_line_mr(view);
```
Note that this function `synth_cz_depth_line_mr` has the same name as the pyfunction. 
Could you please change this name of the rust function to `synth_cz_depth_line_mr_inner` as we did in other synthesis functions?

ShellyGarion commented on crates/accelerate/src/synthesis/linear/lnn.rs:
```diff
@@ -0,0 +1,217 @@
+}
+
+// A single layer of CX gates.
+fn _append_cx_stage1(gates: &mut LnnGatesVec, n: usize) {
```
I think that this function  `_append_cx_stage1` as well as  `_append_cx_stage2` should belong to 
`crates/accelerate/src/synthesis/permutation`
since it's part of the [reverse permutation code](https://github.com/Qiskit/qiskit/blob/main/qiskit/synthesis/permutation/permutation_reverse_lnn.py).

Could you please also port `_append_reverse_permutation_lnn_kms` to rust (as it's quite immediate) ?

ShellyGarion reviewed (commented):
@jlofti - thank you very much for your contribution to Qiskit!
I have a few preliminary comments, about the code structure and performance.

First, could you please transfer your code to a new directory called:
`crates/accelerate/src/synthesis/linear_phase`
(as CZ circuits are not linear). 

Note that there is another algorithm to synthesize linear circuits for LNN (see also https://github.com/Qiskit/qiskit/issues/12228)

In addition, could you perhaps compare the performance of the new rust and original python code for CZ circuits of different sizes ?

Cryoris commented on crates/accelerate/src/synthesis/linear/lnn.rs:
```diff
@@ -0,0 +1,217 @@
+// Appends correct phase gate during CZ synthesis
+fn _append_phase_gate(pat_val: usize, gates: &mut LnnGatesVec, qubit: usize) {
+    // Add phase gates: s, sdg or z
+    if pat_val % 4 == 1 {
```
Could we write this a bit more concisely by using a `match` statement to define the standard gate?
```rust
let gate_id = pat_val % 4;
if gate_id != 0 {
    let gate = match gate_id {
        1 => StandardGate::SdgGate,
        // cover 2 and 3
        _ => unreachable!(),  // unreachable as we have modulo 4
    }
    gates.push((gate, smallvec![], smallvec![Qubit(qubit as u32)];
}
```

Cryoris commented on crates/accelerate/src/synthesis/linear/lnn.rs:
```diff
@@ -0,0 +1,217 @@
+            let pat_val = pats[&(i as isize, j as isize)];
+            if patlist.contains(&pat_val) {
+                let patcnt = patlist.iter().filter(|val| **val == pat_val).count();
+                s_gates[[j]] += patcnt; // qc.sdg[j]
```
If I understand `patcnt` is either 0 or 1 -- could you add a comment explaining that this checks whether an Sdg gate must be added? I think future devs would be grateful (in the Python code it was a bit more obvious IMO) 🙂

Cryoris commented on crates/accelerate/src/synthesis/linear/lnn.rs:
```diff
@@ -0,0 +1,217 @@
+    // s_gates[i] = 0, 1, 2 or 3 for a gate id, sdg, z or s on qubit i respectively
+    let mut s_gates = Array1::<usize>::zeros(num_qubits);
+
+    let mut patlist: Vec<(isize, isize)> = Vec::new();
```
Since all numbers are always positive, would it simplify the code by choosing `usize` as type throughout? That would remove the need for all the `as isize` casts below. Or would that make the logic in the pattern creation tedious?

jlofti replied:
Good catch! I got an subtraction with overflow on initial testing but I guess it was unrelated. Should be fixed

Cryoris reviewed (commented):
Thanks for the contribution @jlofti! Could you maybe run a small benchmark comparing the runtime of the old Python code and the new Rust implementation? It can be something a simple as taking a random binary matrix (rather large, maybe 1000x1000?) and running both codes from the Python interface 🙂

jlofti committed:
added _inner to function name

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
changed directory structure

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
port _append_reverse_permutation_lnn_kms

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
concise function, added sdg comment, all to usize

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti commented:
Hi @ShellyGarion @Cryoris ,

Thanks for all the feedback! Made all the changes, hopefully everything will come back good in CI/CD. My machine was having trouble but for a binary matrix of 350x350 I get the results:

Rust: 308s
Python: 300s

Please let me know if everything looks good or if more changes should be made!

ShellyGarion reviewed (commented):
Thanks @jlofti for handling our comments. 

- Usually, the rust code should be much faster than python. Could you perhaps check why this is not the case here?

- Could you please update the current [reverse permutation code](https://github.com/Qiskit/qiskit/blob/main/qiskit/synthesis/permutation/permutation_reverse_lnn.py) to use rust code?

- Could you please add release notes stating that the CZ synthesis code and reverse permutation code for LNN were ported to rust (you can see examples in previous PRs) ?

alexanderivrii commented:
Hi @jlofti, when you compile the rust code, do you compile it in the release mode? 

```
python setup.py build_rust --inplace --release
```

jlofti committed:
port synth_permutation_reverse_lnn_kms to rust

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
cleanup

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
readded

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti committed:
release notes and simplified comment

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti commented:
> Thanks @jlofti for handling our comments.
> 
>     * Usually, the rust code should be much faster than python. Could you perhaps check why this is not the case here?
> 
>     * Could you please update the current [reverse permutation code](https://github.com/Qiskit/qiskit/blob/main/qiskit/synthesis/permutation/permutation_reverse_lnn.py) to use rust code?
> 
>     * Could you please add release notes stating that the CZ synthesis code and reverse permutation code for LNN were ported to rust (you can see examples in previous PRs) ?



> Hi @jlofti, when you compile the rust code, do you compile it in the release mode?
> 
> ```
> python setup.py build_rust --inplace --release
> ```

Thats exactly what happened lol. Redid the benchmarking and Rust takes about 10s. So seeing around 30x speedup. Also added release notes and changed `synth_permutation_reverse_lnn_kms` to use Rust implementation

alexanderivrii commented on releasenotes/notes/port-synth-cz-depth-line-mr-to-rust-1376d5a41948112a.yaml:
```diff
@@ -0,0 +1,8 @@
+    based on Maslov and Roetteler. On a 350x350 binary matrix, the Rust implementation yields a speedup of about 30 times.
+
+    Additionally, port :func: `.synth_permutation_reverse_lnn_kms` to Rust, which synthesizes a reverse permutation for linear nearest-neighbor architectures using
+    Kutin, Moulton, Smithline method.
```
```suggestion
  - |
    Port :func: `.synth_cz_depth_line_mr` to Rust. This function synthesizes a CZ circuit for linear nearest neighbor (LNN) connectivity, based on the Maslov and Roetteler method. On a 350x350 binary matrix, the Rust implementation yields a speedup of about 30 times.
  - |
    Port :func: `.synth_permutation_reverse_lnn_kms` to Rust, which synthesizes a reverse permutation for linear nearest-neighbor architecture using the Kutin, Moulton, Smithline method.
```

alexanderivrii commented on crates/accelerate/src/synthesis/linear_phase/mod.rs:
```diff
@@ -0,0 +1,35 @@
+
+pub(crate) mod cz_depth_lnn;
+
+#[pyfunction]
+#[pyo3(signature = (mat))]
+fn synth_cz_depth_line_mr(py: Python, mat: PyReadonlyArray2<bool>) -> PyResult<CircuitData> {
```
Could you please add a docstring for this function? Note that the docstrings are prefixed by `///`.

alexanderivrii commented on crates/accelerate/src/synthesis/permutation/mod.rs:
```diff
@@ -114,11 +116,71 @@ pub fn _synth_permutation_depth_lnn_kms(
+    }
+}
+
+#[pyfunction]
+#[pyo3(signature = (num_qubits))]
```
Could you please add a docstring for this function?

alexanderivrii commented on crates/accelerate/src/synthesis/linear_phase/cz_depth_lnn.rs:
```diff
@@ -0,0 +1,172 @@
+}
+
+// Synthesis of a CZ circuit for linear nearest neighbor (LNN) connectivity,
+// based on Maslov and Roetteler.
```
A general comment. Usually we use triple slashes (i.e. `///`) for function docstrings and double slashes (i.e. `//`) for regular comments.  I believe that it makes sense to promote this comment here to a docstring. The same comment applies in many other places.

alexanderivrii reviewed (commented):
The PR is in a great shape. I had a minor nitpick about promoting comments to docstrings  and some suggestions for the release notes, but other than that it looks ready. The 30x improvement in performance is nice :).

jlofti committed:
Update releasenotes/notes/port-synth-cz-depth-line-mr-to-rust-1376d5a41948112a.yaml

Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

jlofti committed:
promoted to docstring and added new docstrings

Signed-off-by: jlofti <65878716+jlofti@users.noreply.github.com>

jlofti commented:
> The PR is in a great shape. I had a minor nitpick about promoting comments to docstrings and some suggestions for the release notes, but other than that it looks ready. The 30x improvement in performance is nice :).

Thanks! Promoted the comments to docstrings and added the ones you requested 👍

jlofti committed:
Merge branch 'main' into port-synth-cz-depth-line-mr

Cryoris commented on crates/accelerate/src/synthesis/linear_phase/cz_depth_lnn.rs:
```diff
@@ -0,0 +1,172 @@
+            1 => StandardGate::SdgGate,
+            2 => StandardGate::ZGate,
+            3 => StandardGate::SGate,
+            // cover 2 and 3
```
Thanks for implementing the change, here's a dangling comment left 🙂 
```suggestion
```

jlofti committed:
Update crates/accelerate/src/synthesis/linear_phase/cz_depth_lnn.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

jlofti commented:
@ShellyGarion @Cryoris @alexanderivrii 

Hi all, was looking for guidance on how to best deduce whats happening with the mac python 3.12 unit tests failing now? from the commit `release notes and simplified comments` to `promoted to docstring and added new docstrings` no code changes were made (just comments). Scratching my head a little trying to track down why these tests are failing now in the qft tests. Looks numpy related?

ShellyGarion commented:
@jlofti - it seems that the failing tests are not related to your PR. Qiskit team in handling it.

jlofti commented:
> @jlofti - it seems that the failing tests are not related to your PR. Qiskit team in handling it.

Ok cool thanks! Just wanted to double check. Took care of all comments so standing by

ShellyGarion commented:
The CI failures should be fixed after https://github.com/Qiskit/qiskit/pull/12979 will be merged

jlofti committed:
Merge branch 'main' into port-synth-cz-depth-line-mr

alexanderivrii reviewed (approved):
LGTM! Thanks for addressing our comments.
